### DISCUSSION TITLE
Default Interface Methods Improvements

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#6171] [Language Ideas]
Time and time again there have been complaints about how DIMs behave in certain scenarios, and how they don't completely solve the problem they were designed for. One notable annoying aspect is **default interface methods are not actually inherited by the implementing types**.

## Motivation through example scenario
Consider an interface `IBlock`, that abstracts away basic properties and methods of blocks (whatever suits your application). Then, you have the implementing structs `DataBlock` and `HeaderBlock`, that both implement the `IBlock` interface. You want to use a great protected method for some instance properties, call it `SomethingAwesome()`, like you would if all those types were classes.

You decide that, since this `SomethingAwesome()` method is to be called on `IBlock`-implementing types, you should abstract it away in the `IBlock` interface, which also defines those said instance properties. You've also heard that "default interface methods" feature, and you try to work it out:
```csharp
public interface IBlock
{
    public int ID { get; }

    protected sealed void SomethingAwesome()
    {
        // probably use ID here
    }
}
```

When you try to call it in your implementing types, you get a weird error you were not expecting, as you were hoping that abstraction of a default method would work like a charm:
```csharp
public struct DataBlock : IBlock
{
    public int ID { get; }

    public void UsingInterfaceMethod()
    {
        (this as IBlock).SomethingAwesome();
    }
}
```

Not only do you have to explicitly cast the `this` instance into its interface type, with your only alternative being `((IBlock)this).SomethingAwesome()`, you also cannot even access the method. This feels highly counter-intuitive, as `protected` implies being visible from inheriting types.

Moreover, there is no option to tell the compiler to include that `SomethingAwesome()` method in your type, unless you explicitly define it, where you will have to call the base method. And if it's `sealed`, you cannot even do that.

## Improvements
The inconveniences are visible: 
- cannot properly access `protected` DIM from implementing types (structs or classes)
- cannot opt-in to including the method to the implementing type, as would happen in the case of classes

Both improvements should enable the following code to actually work:
```csharp
public interface IBlock
{
    protected default sealed void SomethingAwesome() { }
}
public struct DataBlock : IBlock
{
    public void UsingInterfaceMethod()
    {
        SomethingAwesome(); // SomethingAwesome() is now a proud member of DataBlock
    }
}
```

The `default` keyword here signifies that the default-implemented interface method is to be properly inherited in all its implementing types (collisions will cause compiler errors, demanding that the user explicitly override the default interface methods).

Additionally, this would mean the following code would also work:
```csharp
public interface ISomething
{
    public default void NiceMethod() { }
}
public struct StructUsingSomething : ISomething { }

public class UsingStruct
{
    public void TakeUseStruct(StructUsingSomething str)
    {
        str.NiceMethod(); // NiceMethod() is a properly-defined member of StructUsingSomething, directly inherited from ISomething
    }
}
```

**Comment by CyrusNajmabadi** [comment]
> as protected implies being visible from inheriting types.

A core issue is with this.  Classes have never inherited interfaces.  They have only implemented them.  DIM isn't a system to add class-based inheritance to interfaces.  It's a system to provide default implementations such that implementations don't need to provide those implementations themselves.  

It's unclear why you need a DIM here.  Why not just have some sort of helper static method that can do this?  As the method cannot have or access any state, it doesn't need internal access to the type.  So it can be written elsewhere and used by *anything* that wants to consume the actual interface (including subclasses, or anyone else).

**Reply by Rekkonnect** [reply]
> It's unclear why you need a DIM here.

I should have probably further clarified that I want the method to not be static, and rely on some other properties the interface would define. Will edit the original example to reflect those intentions.

**Reply by CyrusNajmabadi** [reply]
a static extension method could still do this.  i'm not saying you're only accessing static properties, simply that you can build this functionality *outside* of the type (e.g. like how static methods work).

**Comment by HaloFour** [comment]
> One notable annoying aspect is default interface methods are not actually inherited by the implementing types.

As suggested by Cyrus, interface members have _never_ been inherited by the implementing type.  The language and runtime allow you to use accessible members that match the signature of the interface members automatically, but the runtime doesn't care how those implementing member slots are filled, which is why explicit implementation exists in C# (and with VB.NET it's much more flexible).  The only guarantee that an interface provides is that you may invoke all of the members _though_ the interface, not through any implementing type.

**Reply by Rekkonnect** [reply]
> interface members have _never_ been inherited by the implementing type

That's a fact, and this proposes change. We're trying to be open to change, and minimize (nullify if possible) the breaking behavior.

> The language and runtime allow you to use accessible members that match the signature of the interface members automatically, but the runtime doesn't care how those implementing member slots are filled, which is why explicit implementation exists in C# (and with VB.NET it's much more flexible).

I don't understand what tha part tries to convey, are you talking about potentially breaking behavior?

> The only guarantee that an interface provides is that you may invoke all of the members though the interface, not through any implementing type.

And that will remain true, inheriting default interface methods will be optional and completely up to the user, which will only enable implementing types to call interface methods that would otherwise be invisible to that type.

**Reply by HaloFour** [reply]
> The `default` keyword here signifies that the default-implemented interface method is to be properly inherited in all its implementing types (collisions will cause compiler errors, demanding that the user explicitly override the default interface methods).

This would mean that adding a `default` DIM member would itself be a breaking change, as you can't know what types may already have a member declared with the same name and signature.  It's likely buried in one of the many comments on this subject as it was being designed but I recall that this is the reason the language team rejected the idea of having the members be visible on the implementing type.

> I don't understand what tha part tries to convey, are you talking about potentially breaking behavior?

That the interface members are never expected to be visible through the implementing type, only the interface.

> And that will remain true, inheriting default interface methods will be optional and completely up to the user, which will only enable implementing types to call interface methods that would otherwise be invisible to that type.

An interface is a contract between the interface and the consumer of that interface, not with the implementers of the interface.

**Reply by Rekkonnect** [reply]
> This would mean that adding a default DIM member would itself be a breaking change, as you can't know what types may already have a member declared with the same name and signature.

`default (virtual)` implies `virtual` but visible to and overridable from the implementor types. Existing signatures would not break, because they would be considered overrides.

> An interface is a contract between the interface and the consumer of that interface, not with the implementers of the interface.

An interface is an interface between its implementors, including the template that should be followed (properties, events and methods), and nowadays, abstracting some common behavior away. It's an evolution of what interfaces are there for. And it doesn't contradict the already existing means for its existence.

**Reply by georg-jung** [reply]
[Docs "Create mixin functionality with default interface methods"](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/mixins-with-default-interface-methods) suggest to use DIMs to create mixins/[traits](https://en.wikipedia.org/wiki/Trait_(computer_programming)). This trait functionality was also part of the [design proposal for DIMs in C# 8](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/default-interface-methods). With this mindset, why should it only be possible to use DIMs to create _public_ functionality?

My use case looks like this: I want to use traits for creating multiple views with shared logic/functionality. I want to have interfaces like `IFilterableByCalendarWeek` and `IFilterableByRegion`. Now I have multiple views which show different kinds of data but are filterable by the same criteria. Thus, beeing able to access a `protected` DIM from the implementing class would make the trait/mixin capabilities of csharp quite much more powerful, as it allows mixing in not only to-be-publicly-consumed behaviour but also behaviour that the implementor itself can make use of.

What do you think?

> Moreover, there is no option to tell the compiler to include that SomethingAwesome() method in your type, unless you explicitly define it, where you will have to call the base method. And if it's sealed, you cannot even do that.

@AlFasGD Can you post an example how this works? I wasn't able to consume the protected DIM from a consuming class using any way I was able to think of.

**Reply by Rekkonnect** [reply]
Sorry for the late reply, you're absolutely right. Protected DIMs do not work for consumption in a non-interface: [repro](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCeAfAAQGYACWBDLCAY1RIEkBYAKAG8WTOTiSBnVCABtUAE24AWEgFlUCABYB7EQAoAlCVYkAvhy4AHdAorUKY/kNESSABUPHTM+UrUbtuzgaOoTliMF4I6DQIVrZePiIAgv6BwY6KKqoA3O4knva+MUEmoXbeptEB2QjxSgBMainMXGl5ESR+RcG54QVZcbIJAKIAHnqCUNRQCJUsOswsPARlJADCJCAMLOzVXFOSpYmpKzU1yvJQvA1H9KoAdJujq5zjNeskAKq8sADmYRkil6rbqTUA9H85hAYDAjCRgN4IABXfi1VqWdCoACOUKgiKOBAADCR5HQwAoAmQYHIMMNLAgcHo6BAjhASCihNgoBhflx9nJDscGOd3vlRF8qrsuKzOACSAB1YaKKEhLEkRFYYQmKAKGAkBToEgiNH5BgAGTgJBwCihJGowNBIVgyAUAGs6AcjrA5gBiEjKegAcjAJAAnL6zr6AKR8KGI77XNyRsUAEVQvD0ZPBqFe3CIhuGJAA7hrbbwAPype68iKFWImL4/SO3LhigBCMpxHKOqroIgU8ZIlsLkfSfLEBEkJbaTQrnXKairNRrosB0WEOIU6rQ6EwImpMDE1FVvChYFQRck9DOw9EZeKm16/UGw0nkZ2XHGWiAA===)

I'd argue that DIMs behave weirdly and that's why people avoid using them except for very standard scenarios, rarely falling into the niche cases. And then somebody would come in and say "we don't quite like the approach of needing DIMs anyway" or "I wouldn't care too much to have that such a high priority".

**Comment by TahirAhmadov** [comment]
Can you use an `internal interface` to accomplish what you need?
